pkg/datasources/functions.go

```diff
@@ -60,43 +62,58 @@ var functionsSchema = map[string]*schema.Schema{
 
 func Functions() *schema.Resource {
 	return &schema.Resource{
-		Read:   ReadFunctions,
-		Schema: functionsSchema,
+		ReadContext: ReadContextFunctions,
+		Schema:      functionsSchema,
 	}
 }
 
-// todo: fix this. ListUserFunctions isn't using the right struct right now and also the signature of this doesn't support all the features it could for example, database and schema should be optional, and you could also list by account.
-func ReadFunctions(d *schema.ResourceData, meta interface{}) error {
+func ReadContextFunctions(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
 	databaseName := d.Get("database").(string)
 	schemaName := d.Get("schema").(string)
 
-	d.SetId("functions")
-	currentFunctions, err := snowflake.ListUserFunctions(databaseName, schemaName, db)
-	if err != nil {
-		log.Printf("[DEBUG] error listing functions: %v", err)
-		return nil
+	request := sdk.NewShowFunctionRequest()
+	if databaseName != "" {
+		request.WithIn(&sdk.In{Database: sdk.NewAccountObjectIdentifier(databaseName)})
 	}
+	if schemaName != "" {
+		request.WithIn(&sdk.In{Schema: sdk.NewDatabaseObjectIdentifier(databaseName, schemaName)})
+	}
+	functions, err := client.Functions.Show(ctx, request)
+	if err != nil {
+		id := d.Id()
 
-	functions := []map[string]interface{}{}
-
-	for _, function := range currentFunctions {
-		functionMap := map[string]interface{}{}
+		d.SetId("")
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  fmt.Sprintf("Unable to parse functions in schema (%s)", id),
+				// TODO: link to the design decisions doc
+				Detail: "See our document on design decisions for functions: <LINK (coming soon)>",
```

Comment:
I don't think we're going to release any design decisions on functions, are we? (this looks like copy-paste error with grants)

Reply:
removed